XMonad/Util/Loggers.hs

```diff
@@ -251,6 +254,70 @@ instance Default TitlesFormat where
+logClassnamesOnScreen
+  :: ScreenId           -- ^ Screen to log the classnames on
+  -> (String -> String) -- ^ Formatting for the focused   window
+  -> (String -> String) -- ^ Formatting for the unfocused window
+  -> Logger
+logClassnamesOnScreen sid formatFoc formatUnfoc =
+  logClassnamesOnScreen' sid ClassnamesFormat{ focusedFormatClassname   = formatFoc
+                                             , unfocusedFormatClassname = formatUnfoc
+                                             , urgentFormatClassname    = formatUnfoc
+                                             }
+
+-- | Like 'logClassnamesOnScreen' but with support for urgent windows.  To
+-- be used with "XMonad.Hooks.UrgencyHook".
+logClassnamesOnScreen' :: ScreenId -> ClassnamesFormat -> Logger
+logClassnamesOnScreen' sid (ClassnamesFormat formatFoc formatUnfoc formatUrg) =
+  (`withScreen` sid) $ \screen -> do
+    let focWin = fmap W.focus . W.stack . W.workspace $ screen
+    urgWins <- readUrgents
+    logClassnamesOnScreenWorker screen $ \win name ->
+      if | Just win == focWin -> formatFoc   name
+         | win `elem` urgWins -> formatUrg   name
+         | otherwise          -> formatUnfoc name
+
+-- | Internal function for 'logClassnamesOnScreen' and 'logClassnamesOnScreen''.
+logClassnamesOnScreenWorker :: WindowScreen -> (Window -> String -> String) -> Logger
+logClassnamesOnScreenWorker screen logger = do
+  let wins = maybe [] W.integrate . W.stack . W.workspace $ screen
+  winNames <- traverse (fmap show . getNameWMClass) wins
+  pure . Just . unwords $ zipWith logger wins winNames
+
+-- | Like 'logClassnamesOnScreen', but directly use the "focused" screen
+-- (the one with the currently focused workspace).
+logClassnames :: (String -> String) -> (String -> String) -> Logger
+logClassnames formatFoc formatUnfoc = do
+  sid <- gets $ W.screen . W.current . windowset
+  logClassnamesOnScreen sid formatFoc formatUnfoc
+
+-- | Variant of 'logClassnames', but with support for urgent windows.
+logClassnames' :: ClassnamesFormat -> Logger
+logClassnames' formatter =
+  gets (W.screen . W.current . windowset) >>= (`logClassnamesOnScreen'` formatter)
+
+-- | Formatting applied to the classnames of certain windows.
+data ClassnamesFormat = ClassnamesFormat
+  { focusedFormatClassname   :: String -> String  -- ^ Focused formatting.
+  , unfocusedFormatClassname :: String -> String  -- ^ Unfocused formatting.
+  , urgentFormatClassname    :: String -> String  -- ^ Formatting when urgent.
+  }
+
+-- | How to format these classnames by default when using 'logClassnames'' and
+-- 'logClassnamesOnScreen''.
+instance Default ClassnamesFormat where
+  def = ClassnamesFormat
+    { focusedFormatClassname   = wrap "[" "]" . xmobarRaw . shorten 30 . xmobarStrip
+    , unfocusedFormatClassname =                xmobarRaw . shorten 30 . xmobarStrip
+    , urgentFormatClassname    = wrap "!" "!" . xmobarRaw . shorten 30 . xmobarStrip
+    }
+
```

Comment:
Similar to what @slotThe, maybe we use the same record `TitlesFormat` for this as well? Maybe

```
type ClassnamesFormat = TitlesFormat
```

Reply:
My only worry with this is it leaves the constructors feeling a bit weird. I.e., building a `ClassnamesFormat` with a function `TitlesFormat` feels off to me. I don't consider myself a god of haskell, is there a neat way to get around this? Or, I'm happy to go with it anyway, if you feel it won't be an issue?

Reply:
That's a valid concern. I'm not sure how big of an issue this actually is though:
- People use `def` in their configs
- There's a single function that's explicitly "deconstructing"  `ClassnamesFormat`

Let's see what others think though, I don't feel strongly about it.

Reply:
Perhaps just refactor a bit?
```
xmobarWsFormat, xmobarFocusedFormat, xmobarUrgentFormat :: String -> String
xmobarWsFormat = xmobarRaw . shorten 30 . xmobarStrip
xmobarFocusedFormat = wrap "[" "]" . xmobarWsFormat
xmobarUrgentFormat = wrap "!" "!" . xmobarWsFormat
```